app/multitenant/memcache_client.go

```diff
@@ -0,0 +1,131 @@
+package multitenant
+
+import (
+	"bytes"
+	"fmt"
+	"net"
+	"sort"
+	"time"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/bradfitz/gomemcache/memcache"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/weaveworks/scope/report"
+)
+
+var (
+	memcacheCounter = prometheus.NewCounterVec(prometheus.CounterOpts{
+		Namespace: "scope",
+		Name:      "memcache",
+		Help:      "Reports that missed our in-memory cache but went to our memcache",
+	}, []string{"result"})
+
+	memcacheRequestDuration = prometheus.NewSummaryVec(prometheus.SummaryOpts{
+		Namespace: "scope",
+		Name:      "memcache_request_duration_nanoseconds",
+		Help:      "Time spent doing memcache requests.",
+	}, []string{"method", "status_code"})
+)
+
+func init() {
+	prometheus.MustRegister(memcacheCounter)
+	prometheus.MustRegister(memcacheRequestDuration)
+}
+
+type memcacheClient struct {
+	client     *memcache.Client
+	expiration int32
+}
+
+// Create a new MemcacheClient that gets its server list from SRV and updates
+// the server list on a regular basis.
+func NewMemcacheClient(host string, timeout time.Duration, service string, updateInterval time.Duration, expiration int32) (*memcacheClient, error) {
+	var servers memcache.ServerList
+	client := memcache.NewFromSelector(&servers)
+	client.Timeout = timeout
+	err := updateMemcacheServers(host, service, &servers)
+	if err != nil {
+		return nil, err
+	}
+
+	ticker := time.NewTicker(updateInterval)
+	go func() {
+		for range ticker.C {
+			updateMemcacheServers(host, service, &servers)
+		}
+	}()
+	return &memcacheClient{client, expiration}, nil
+}
+
+// updateMemcacheServers sets a memcache server list from SRV records. SRV
+// priority & weight are ignored.
+func updateMemcacheServers(hostname, service string, serverList *memcache.ServerList) error {
+	_, addrs, err := net.LookupSRV(service, "tcp", hostname)
+	if err != nil {
+		return err
+	}
+	var servers []string
+	for _, srv := range addrs {
+		servers = append(servers, fmt.Sprintf("%s:%d", srv.Target, srv.Port))
+	}
+	// ServerList deterministically maps keys to _index_ of the server list.
+	// Since DNS returns records in different order each time, we sort to
+	// guarantee best possible match between nodes.
+	sort.Strings(servers)
+	return serverList.SetServers(servers...)
+}
+
+func (c memcacheClient) FetchReports(keys []string) ([]report.Report, []string, error) {
```

Comment:
I don't know. I made it a value method because it doesn't change any state (following the advice in the [go FAQ](https://golang.org/doc/faq#methods_on_values_or_pointers)).